### DISCUSSION TITLE
How to get the context type from a specific procedure?

### DISCUSSION THREAD

**Discussion by dBianchii** [trpc/trpc#5472] [Q&A] [answered]
I am trying to infer the context type from a specific procedure. I have one specific `isTeamOwnerProcedure` which extends the context:

```ts
export const isTeamOwnerProcedure = protectedProcedure.use(
  async ({ ctx, next }) => {
    const team = await ctx.prisma.team.findUniqueOrThrow({
      where: {
        id: ctx.session.user.activeTeamId,
      },
      select: {
        ownerId: true,
      },
    });

    if (team.ownerId !== ctx.session.user.id)
      throw new TRPCError({
        message: "Only the team owner can do this",
        code: "FORBIDDEN",
      });

    return next({
      ctx: {
        ...ctx,
        team,
      },
    });
  },
);
```

I then want to get the type of the context from this procedure, but couldn't find a way to do it. I found a helper which infers from a router, like this:

```ts
type RootCtx = inferRouterContext<AppRouter>
```

But I can't do anything much from this. I can't seem to be able to do for example `type GetAllAppsCtx = inferRouterContext<AppRouter["app"]["getAll"]>`

Is there a way for me to get the context from a specific procedure, or from a specific router at the end of one of my AppRouter branches? (like we can do with `RouterOutputs['post']['all'] `for ex.)

**Comment by KATT** [comment]
Yeah, it'd be nice to do something like `inferProcedureBuilder<typeof isTeamOwnerProcedure>` here & get

```ts
type Types = inferProcedureBuilder<typeof isTeamOwnerProcedure>
```

where `Types` would be something like

```ts
type Types = {
  ctx: Merge<BaseContext, team: Team }>
}
```

**Reply by KATT** [reply]
Feel free to open a feature request

An ugly workaround could be

```ts
type TeamOwnerProcedure = typeof isTeamOwnerProcedure 
type TeamOwnerProcedureOptions = Parameters<Parameters<TeamOwnerProcedure['query']>[0]>[0]
type TeamOwnerProcedureContext = TeamOwnerProcedureOptions['ctx']
```

**Reply by dBianchii** [reply]
I saw in cal.com's codebase so many manually defined contexts for all of their api handlers. This becomes hard to extend when your contexts get more complex

Example: https://github.com/calcom/cal.com/blob/main/packages/trpc/server/routers/loggedInViewer/appById.handler.ts#L9

I'll open a feature request

**Reply by nordowl** [reply]
Faced the same problem. KATTs workaround works great!
I made myself a little helper type to more easily get multiple context types.

```typescript
import { ProcedureBuilder } from "@trpc/server/dist/core/internals/procedureBuilder"

export const publicProcedure = t.procedure.use(allowPublic)

type ExtractProcedureContext<T extends ProcedureBuilder<any>> = Parameters<Parameters<T["query"]>[0]>[0]["ctx"]

export type PublicProcedureContext = ExtractProcedureContext<typeof publicProcedure>
```

**Reply by dBianchii** [reply]
> Faced the same problem. KATTs workaround works great!
> 
> I made myself a little helper type to more easily get multiple context types.
> 
> 
> 
> ```typescript
> 
> import { ProcedureBuilder } from "@trpc/server/dist/core/internals/procedureBuilder"
> 
> 
> 
> export const publicProcedure = t.procedure.use(allowPublic)
> 
> 
> 
> type ExtractProcedureContext<T extends ProcedureBuilder<any>> = Parameters<Parameters<T["query"]>[0]>[0]["ctx"]
> 
> 
> 
> export type PublicProcedureContext = ExtractProcedureContext<typeof publicProcedure>
> 
> ```

If you want you can try to use the helper Katt provided on the latest version of the v11 beta

**Reply by KATT** [reply, marked as answer]
https://trpc.io/docs/server/procedures#inferProcedureBuilderResolverOptions
